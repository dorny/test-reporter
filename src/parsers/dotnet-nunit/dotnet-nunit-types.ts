export interface NunitReport {
  'test-run': TestRun
}

export interface TestRun {
  $: {
    time: string
  }
  'test-suite'?: TestSuite[]
}

export interface TestSuite {
  $: {
    name: string
    tests: string
    errors: string
    failed: string
    skipped: string
    passed: string
    duration: string
  }
  'test-case'?: TestCase[]
  'test-suite'?: TestSuite[]
}

export interface TestCase {
  $: {
    fullname: string
    name: string
    time: string
    result: string
  }
  failure?: Failure[]
}

export interface Failure {
  'stack-trace'?: string
  message: string
}
