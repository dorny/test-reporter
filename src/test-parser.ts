import {TestRunResult} from './test-results'

export interface ParseOptions {
  parseErrors: boolean
  workDir?: string
  trackedFiles: string[]
  directoryMapping?: {from: string; to: string}
}

export interface TestParser {
  parse(path: string, content: string): Promise<TestRunResult>
}
