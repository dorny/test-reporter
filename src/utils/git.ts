export async function listFiles(): Promise<string[]> {
  return []
}
