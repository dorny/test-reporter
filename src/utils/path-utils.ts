export function normalizeDirPath(path: string, addTrailingSlash: boolean): string {
  if (!path) {
    return path
  }

  path = normalizeFilePath(path)
  if (addTrailingSlash && !path.endsWith('/')) {
    path += '/'
  }
  return path
}

export function normalizeFilePath(path: string): string {
  if (!path) {
    return path
  }

  return path.trim().replace(/\\/g, '/')
}

export function getBasePath(path: string, trackedFiles: string[]): string | undefined {
  if (trackedFiles.includes(path)) {
    return ''
  }

  for (const file of trackedFiles) {
    const pathParts = path.split('/')
    const originalLength = pathParts.length
    const fileParts = file.split('/')

    while (pathParts.length && pathParts.slice(-1)[0] === fileParts.slice(-1)[0]) {
      pathParts.pop()
      fileParts.pop()
    }

    // we found some matching path parts
    if (pathParts.length !== originalLength) {
      return pathParts.join('/')
    }
  }

  return undefined
}
