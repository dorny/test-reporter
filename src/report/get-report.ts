import * as core from '@actions/core'
import {TestExecutionResult, TestRunResult, TestSuiteResult} from '../test-results'
import {Align, formatTime, Icon, link, table} from '../utils/markdown-utils'
import {DEFAULT_LOCALE} from '../utils/node-utils'
import {getFirstNonEmptyLine} from '../utils/parse-utils'
import {slug} from '../utils/slugger'

const MAX_REPORT_LENGTH = 65535
const MAX_ACTIONS_SUMMARY_LENGTH = 131072 // 1048576 soon

export interface ReportOptions {
  listSuites: 'all' | 'failed' | 'none'
  listTests: 'all' | 'failed' | 'none'
  baseUrl: string
  onlySummary: boolean
  useActionsSummary: boolean
  badgeTitle: string
}

const defaultOptions: ReportOptions = {
  listSuites: 'all',
  listTests: 'all',
  baseUrl: '',
  onlySummary: false,
  useActionsSummary: true,
  badgeTitle: 'tests'
}

export function getReport(results: TestRunResult[], options: ReportOptions = defaultOptions): string {
  core.info('Generating check run summary')

  applySort(results)

  const opts = {...options}
  let lines = renderReport(results, opts)
  let report = lines.join('\n')

  if (getByteLength(report) <= getMaxReportLength(options)) {
    return report
  }

  if (opts.listTests === 'all') {
    core.info("Test report summary is too big - setting 'listTests' to 'failed'")
    opts.listTests = 'failed'
    lines = renderReport(results, opts)
    report = lines.join('\n')
    if (getByteLength(report) <= getMaxReportLength(options)) {
      return report
    }
  }

  core.warning(`Test report summary exceeded limit of ${getMaxReportLength(options)} bytes and will be trimmed`)
  return trimReport(lines, options)
}

function getMaxReportLength(options: ReportOptions = defaultOptions): number {
  return options.useActionsSummary ? MAX_ACTIONS_SUMMARY_LENGTH : MAX_REPORT_LENGTH
}

function trimReport(lines: string[], options: ReportOptions): string {
  const closingBlock = '```'
  const errorMsg = `**Report exceeded GitHub limit of ${getMaxReportLength(options)} bytes and has been trimmed**`
  const maxErrorMsgLength = closingBlock.length + errorMsg.length + 2
  const maxReportLength = getMaxReportLength(options) - maxErrorMsgLength

  let reportLength = 0
  let codeBlock = false
  let endLineIndex = 0
  for (endLineIndex = 0; endLineIndex < lines.length; endLineIndex++) {
    const line = lines[endLineIndex]
    const lineLength = getByteLength(line)

    reportLength += lineLength + 1
    if (reportLength > maxReportLength) {
      break
    }

    if (line === '```') {
      codeBlock = !codeBlock
    }
  }

  const reportLines = lines.slice(0, endLineIndex)
  if (codeBlock) {
    reportLines.push('```')
  }
  reportLines.push(errorMsg)
  return reportLines.join('\n')
}

function applySort(results: TestRunResult[]): void {
  results.sort((a, b) => a.path.localeCompare(b.path, DEFAULT_LOCALE))
  for (const res of results) {
    res.suites.sort((a, b) => a.name.localeCompare(b.name, DEFAULT_LOCALE))
  }
}

function getByteLength(text: string): number {
  return Buffer.byteLength(text, 'utf8')
}

function renderReport(results: TestRunResult[], options: ReportOptions): string[] {
  const sections: string[] = []
  const badge = getReportBadge(results, options)
  sections.push(badge)

  const runs = getTestRunsReport(results, options)
  sections.push(...runs)

  return sections
}

function getReportBadge(results: TestRunResult[], options: ReportOptions): string {
  const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
  const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
  const failed = results.reduce((sum, tr) => sum + tr.failed, 0)
  return getBadge(passed, failed, skipped, options)
}

function getBadge(passed: number, failed: number, skipped: number, options: ReportOptions): string {
  const text = []
  if (passed > 0) {
    text.push(`${passed} passed`)
  }
  if (failed > 0) {
    text.push(`${failed} failed`)
  }
  if (skipped > 0) {
    text.push(`${skipped} skipped`)
  }
  const message = text.length > 0 ? text.join(', ') : 'none'

  let color = 'success'
  if (failed > 0) {
    color = 'critical'
  } else if (passed === 0 && failed === 0) {
    color = 'yellow'
  }
  const hint = failed > 0 ? 'Tests failed' : 'Tests passed successfully'
  const uri = encodeURIComponent(`${options.badgeTitle}-${message}-${color}`)
  return `![${hint}](https://img.shields.io/badge/${uri})`
}

function getTestRunsReport(testRuns: TestRunResult[], options: ReportOptions): string[] {
  const sections: string[] = []
  const totalFailed = testRuns.reduce((sum, tr) => sum + tr.failed, 0)
  if (totalFailed === 0) {
    sections.push(`<details><summary>Expand for details</summary>`)
    sections.push(` `)
  }

  const shouldShowCoverage = testRuns.some(tr => tr.coverage !== undefined)

  const columnNames = ['Report', 'Passed', 'Failed', 'Skipped', 'Time']
  const columnAligns = [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right]

  if (shouldShowCoverage) {
    columnNames.push('Coverage')
    columnAligns.push(Align.Right)
  }

  if (testRuns.length > 0 || options.onlySummary) {
    const tableData = testRuns
      .filter(tr => tr.passed > 0 || tr.failed > 0 || tr.skipped > 0)
      .map(tr => {
        const time = formatTime(tr.time)
        const name = tr.path
        const passed = tr.passed > 0 ? `${tr.passed} ${Icon.success}` : ''
        const failed = tr.failed > 0 ? `${tr.failed} ${Icon.fail}` : ''
        const skipped = tr.skipped > 0 ? `${tr.skipped} ${Icon.skip}` : ''

        if (shouldShowCoverage) {
          const coverage = tr.coverage !== undefined ? tr.coverage + '%' : ''
          return [name, passed, failed, skipped, time, coverage]
        }

        return [name, passed, failed, skipped, time]
      })

    const resultsTable = table(columnNames, columnAligns, ...tableData)
    sections.push(resultsTable)
  }

  if (options.onlySummary === false) {
    const suitesReports = testRuns.map((tr, i) => getSuitesReport(tr, i, options)).flat()
    sections.push(...suitesReports)
  }

  if (totalFailed === 0) {
    sections.push(`</details>`)
  }
  return sections
}

function getSuitesReport(tr: TestRunResult, runIndex: number, options: ReportOptions): string[] {
  const sections: string[] = []
  const suites = options.listSuites === 'failed' ? tr.failedSuites : tr.suites

  if (options.listSuites !== 'none') {
    const trSlug = makeRunSlug(runIndex)
    const nameLink = `<a id="${trSlug.id}" href="${options.baseUrl + trSlug.link}">${tr.path}</a>`
    const icon = getResultIcon(tr.result)
    sections.push(`## ${icon}\xa0${nameLink}`)

    const time = formatTime(tr.time)
    const headingLine2 =
      tr.tests > 0
        ? `**${tr.tests}** tests were completed in **${time}** with **${tr.passed}** passed, **${tr.failed}** failed and **${tr.skipped}** skipped.`
        : 'No tests found'
    sections.push(headingLine2)

    if (suites.length > 0) {
      const shouldShowCoverage = suites.some(s => s.coverage !== undefined)

      const columnNames = ['Test suite', 'Passed', 'Failed', 'Skipped', 'Time']
      const columnAligns = [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right]

      if (shouldShowCoverage) {
        columnNames.push('Coverage')
        columnAligns.push(Align.Right)
      }

      const suitesTable = table(
        columnNames,
        columnAligns,
        ...suites.map((s, suiteIndex) => {
          const tsTime = formatTime(s.time)
          const tsName = s.name
          const skipLink = options.listTests === 'none' || (options.listTests === 'failed' && s.result !== 'failed')
          const tsAddr = options.baseUrl + makeSuiteSlug(runIndex, suiteIndex).link
          const tsNameLink = skipLink ? tsName : link(tsName, tsAddr)
          const passed = s.passed > 0 ? `${s.passed} ${Icon.success}` : ''
          const failed = s.failed > 0 ? `${s.failed} ${Icon.fail}` : ''
          const skipped = s.skipped > 0 ? `${s.skipped} ${Icon.skip}` : ''

          const coverage = s.coverage !== undefined ? s.coverage + '%' : ''

          if (shouldShowCoverage) {
            return [tsNameLink, passed, failed, skipped, tsTime, coverage]
          }

          return [tsNameLink, passed, failed, skipped, tsTime]
        })
      )
      sections.push(suitesTable)
    }
  }

  if (options.listTests !== 'none') {
    const tests = suites.map((ts, suiteIndex) => getTestsReport(ts, runIndex, suiteIndex, options)).flat()

    if (tests.length > 1) {
      sections.push(...tests)
    }
  }

  return sections
}

function getTestsReport(ts: TestSuiteResult, runIndex: number, suiteIndex: number, options: ReportOptions): string[] {
  if (options.listTests === 'failed' && ts.result !== 'failed') {
    return []
  }
  const groups = ts.groups
  if (groups.length === 0) {
    return []
  }

  const sections: string[] = []

  const tsName = ts.name
  const tsSlug = makeSuiteSlug(runIndex, suiteIndex)
  const tsNameLink = `<a id="${tsSlug.id}" href="${options.baseUrl + tsSlug.link}">${tsName}</a>`
  const icon = getResultIcon(ts.result)
  sections.push(`### ${icon}\xa0${tsNameLink}`)

  sections.push('```')
  for (const grp of groups) {
    if (grp.name) {
      sections.push(grp.name)
    }
    const space = grp.name ? '  ' : ''
    for (const tc of grp.tests) {
      const result = getResultIcon(tc.result)
      sections.push(`${space}${result} ${tc.name}`)
      if (tc.error) {
        const lines = (tc.error.message ?? getFirstNonEmptyLine(tc.error.details)?.trim())
          ?.split(/\r?\n/g)
          .map(l => '\t' + l)
        if (lines) {
          sections.push(...lines)
        }
      }
    }
  }
  sections.push('```')

  return sections
}

function makeRunSlug(runIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}`)
}

function makeSuiteSlug(runIndex: number, suiteIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}s${suiteIndex}`)
}

function getResultIcon(result: TestExecutionResult): string {
  switch (result) {
    case 'success':
      return Icon.success
    case 'skipped':
      return Icon.skip
    case 'failed':
      return Icon.fail
    default:
      return ''
  }
}
