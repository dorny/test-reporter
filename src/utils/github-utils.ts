import {createWriteStream} from 'fs'
import * as core from '@actions/core'
import * as github from '@actions/github'
import {GitHub} from '@actions/github/lib/utils'
import * as stream from 'stream'
import {promisify} from 'util'
import got from 'got'
const asyncStream = promisify(stream.pipeline)

export function getCheckRunContext(): {sha: string; runId: number; branch: string} {
  let branch = github.context.ref
  if (branch.startsWith('refs/heads')) branch = branch.slice(11)
  core.info('Branch: ' + branch)

  if (github.context.eventName === 'workflow_run') {
    core.info('Action was triggered by workflow_run: using SHA and RUN_ID from triggering workflow')
    const event = github.context.payload
    if (!event.workflow_run) {
      throw new Error("Event of type 'workflow_run' is missing 'workflow_run' field")
    }
    return {
      sha: event.workflow_run.head_commit.id,
      runId: event.workflow_run.id,
      branch
    }
  }

  const runId = github.context.runId
  if (github.context.payload.pull_request) {
    core.info(`Action was triggered by ${github.context.eventName}: using SHA from head of source branch`)
    const pr = github.context.payload.pull_request
    return {sha: pr.head.sha, runId, branch}
  }

  return {sha: github.context.sha, runId, branch}
}

export async function downloadArtifact(
  octokit: InstanceType<typeof GitHub>,
  artifactId: number,
  fileName: string,
  token: string
): Promise<void> {
  core.startGroup(`Downloading artifact ${fileName}`)
  try {
    core.info(`Artifact ID: ${artifactId}`)

    const req = octokit.rest.actions.downloadArtifact.endpoint({
      ...github.context.repo,
      artifact_id: artifactId,
      archive_format: 'zip'
    })

    const headers = {
      Authorization: `Bearer ${token}`
    }
    const resp = await got(req.url, {
      headers,
      followRedirect: false
    })

    core.info(`Fetch artifact URL: ${resp.statusCode} ${resp.statusMessage}`)
    if (resp.statusCode !== 302) {
      throw new Error('Fetch artifact URL failed: received unexpected status code')
    }

    const url = resp.headers.location
    if (url === undefined) {
      const receivedHeaders = Object.keys(resp.headers)
      core.info(`Received headers: ${receivedHeaders.join(', ')}`)
      throw new Error('Location header was not found in API response')
    }
    if (typeof url !== 'string') {
      throw new Error(`Location header has unexpected value: ${url}`)
    }

    const downloadStream = got.stream(url, {headers})
    const fileWriterStream = createWriteStream(fileName)

    core.info(`Downloading ${url}`)
    downloadStream.on('downloadProgress', ({transferred}) => {
      core.info(`Progress: ${transferred} B`)
    })
    await asyncStream(downloadStream, fileWriterStream)
  } finally {
    core.endGroup()
  }
}

export async function listFiles(octokit: InstanceType<typeof GitHub>, sha: string): Promise<string[]> {
  core.startGroup('Fetching list of tracked files from GitHub')
  try {
    const commit = await octokit.rest.git.getCommit({
      commit_sha: sha,
      ...github.context.repo
    })
    const files = await listGitTree(octokit, commit.data.tree.sha, '')
    return files
  } finally {
    core.endGroup()
  }
}

async function listGitTree(octokit: InstanceType<typeof GitHub>, sha: string, path: string): Promise<string[]> {
  const pathLog = path ? ` at ${path}` : ''
  core.info(`Fetching tree ${sha}${pathLog}`)
  let truncated = false
  let tree = await octokit.rest.git.getTree({
    recursive: 'true',
    tree_sha: sha,
    ...github.context.repo
  })

  if (tree.data.truncated) {
    truncated = true
    tree = await octokit.rest.git.getTree({
      tree_sha: sha,
      ...github.context.repo
    })
  }

  const result: string[] = []
  for (const tr of tree.data.tree) {
    const file = `${path}${tr.path}`
    if (tr.type === 'blob') {
      result.push(file)
    } else if (tr.type === 'tree' && truncated) {
      const files = await listGitTree(octokit, tr.sha || '', `${file}/`)
      result.push(...files)
    }
  }

  return result
}
