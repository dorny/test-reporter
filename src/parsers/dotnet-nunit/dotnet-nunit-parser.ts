import * as path from 'path'
import {ParseOptions, TestParser} from '../../test-parser'
import {parseStringPromise} from 'xml2js'

import {NunitReport, TestCase, TestSuite} from './dotnet-nunit-types'
import {normalizeFilePath} from '../../utils/path-utils'

import {
  TestExecutionResult,
  TestRunResult,
  TestSuiteResult,
  TestGroupResult,
  TestCaseResult,
  TestCaseError
} from '../../test-results'

export class DotnetNunitParser implements TestParser {
  readonly trackedFiles: {[fileName: string]: string[]}

  constructor(readonly options: ParseOptions) {
    this.trackedFiles = {}
    for (const filePath of options.trackedFiles) {
      const fileName = path.basename(filePath)
      const files = this.trackedFiles[fileName] ?? (this.trackedFiles[fileName] = [])
      files.push(normalizeFilePath(filePath))
    }
  }

  async parse(filePath: string, content: string): Promise<TestRunResult> {
    const reportOrSuite = await this.getNunitReport(filePath, content)
    return this.getTestRunResult(filePath, reportOrSuite)
  }

  private async getNunitReport(filePath: string, content: string): Promise<NunitReport> {
    try {
      return await parseStringPromise(content)
    } catch (e) {
      throw new Error(`Invalid XML at ${filePath}\n\n${e}`)
    }
  }

  private getTestSuiteResultRecursive(
    testSuites: TestSuite[] | undefined,
    suiteResults: TestSuiteResult[],
    depth: number
  ): void {
    if (testSuites !== undefined) {
      testSuites.map(ts => {
        const name = ts.$.name.trim()
        const time = parseFloat(ts.$.duration) * 1000
        const groups = this.getGroups(ts)
        const sr = new TestSuiteResult(name, groups, time, depth)
        suiteResults.push(sr)

        if (groups.length === 0) {
          const nestedTestSuites = ts['test-suite']
          if (nestedTestSuites !== undefined) {
            this.getTestSuiteResultRecursive(nestedTestSuites, suiteResults, depth + 1)
          }
        }
      })
    }
  }

  private getTestRunResult(filePath: string, nunit: NunitReport): TestRunResult {
    const suites: TestSuiteResult[] = []

    const testSuites = nunit['test-run']['test-suite']
    this.getTestSuiteResultRecursive(testSuites, suites, 0)

    const seconds = parseFloat(nunit['test-run'].$?.time)
    const time = isNaN(seconds) ? undefined : seconds * 1000
    return new TestRunResult(filePath, suites, time)
  }

  private getGroups(suite: TestSuite): TestGroupResult[] {
    const groups: {describe: string; tests: TestCase[]}[] = []
    if (suite['test-case'] === undefined) {
      return []
    }
    for (const tc of suite['test-case']) {
      let grp = groups.find(g => g.describe === tc.$.name)
      if (grp === undefined) {
        grp = {describe: tc.$.name, tests: []}
        groups.push(grp)
      }
      grp.tests.push(tc)
    }

    return groups.map(grp => {
      const tests = grp.tests.map(tc => {
        const name = tc.$.name.trim()
        const result = this.getTestCaseResult(tc)
        const time = parseFloat(tc.$.time) * 1000
        const error = this.getTestCaseError(tc)
        return new TestCaseResult(name, result, time, error)
      })
      return new TestGroupResult(grp.describe, tests)
    })
  }

  private getTestCaseResult(test: TestCase): TestExecutionResult {
    if (test.failure) return 'failed'
    if (test.$.result === 'Skipped') return 'skipped'
    return 'success'
  }

  private getTestCaseError(tc: TestCase): TestCaseError | undefined {
    if (!this.options.parseErrors) {
      return undefined
    }

    const failure = tc.failure
    if (!failure) {
      return undefined
    }

    const details = failure[0]['stack-trace'] === undefined ? '' : failure[0]['stack-trace'][0]

    let filePath
    let line

    const src = this.exceptionThrowSource(details)
    if (src) {
      filePath = src.filePath
      line = src.line
    }

    return {
      path: filePath,
      line,
      details,
      message: failure[0].message === undefined ? '' : failure[0].message[0]
    }
  }

  private exceptionThrowSource(stackTrace: string): {filePath: string; line: number} | undefined {
    const lines = stackTrace.split(/\r?\n/)
    const re = /^at (.*\) in .*):(.+)$/

    for (const str of lines) {
      const match = str.match(re)
      if (match !== null) {
        const [, , filePath, lineStr] = match
        const line = parseInt(lineStr)
        return {filePath, line}
      }
    }
  }
}
