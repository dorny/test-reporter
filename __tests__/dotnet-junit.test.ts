import * as fs from 'fs'
import * as path from 'path'

import {DotnetNunitParser as DotnetNunitParser} from '../src/parsers/dotnet-nunit/dotnet-nunit-parser'
import {ParseOptions} from '../src/test-parser'
import {getReport, ReportOptions} from '../src/report/get-report'
import {normalizeFilePath} from '../src/utils/path-utils'

describe('dotnet-nunit tests', () => {
  it('report parses without errors', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'external', 'dotnet', 'nunit-test-results.xml')
    const outputPath = path.join(__dirname, '__outputs__', 'dotnet-nunit-tests-results.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})
    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: []
    }

    const parser = new DotnetNunitParser(opts)
    const result = await parser.parse(filePath, fileContent)

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })

  it('report from matches snapshot', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'external', 'dotnet', 'nunit-test-results.xml')
    const outputPath = path.join(__dirname, '__outputs__', 'dotnet-nunit-tests-results.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: []
    }

    const parser = new DotnetNunitParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result).toMatchSnapshot()

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })
})
