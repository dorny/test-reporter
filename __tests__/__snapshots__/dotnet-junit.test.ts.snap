// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dotnet-nunit tests report from matches snapshot 1`] = `
TestRunResult {
  "path": "fixtures/external/dotnet/nunit-test-results.xml",
  "suites": Array [
    TestSuiteResult {
      "groups": Array [],
      "indentationDepth": 0,
      "name": "mock-assembly.dll",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [
        TestGroupResult {
          "name": "FailingTest",
          "tests": Array [
            TestCaseResult {
              "error": Object {
                "details": "   at NUnit.Framework.Assert.Fail(String message, Object[] args) in D:\\\\Dev\\\\NUnit\\\\nunit-3.0\\\\work\\\\NUnitFramework\\\\src\\\\framework\\\\Assert.cs:line 142
   at NUnit.Framework.Assert.Fail(String message) in D:\\\\Dev\\\\NUnit\\\\nunit-3.0\\\\work\\\\NUnitFramework\\\\src\\\\framework\\\\Assert.cs:line 152
   at NUnit.Tests.Assemblies.MockTestFixture.FailingTest() in D:\\\\Dev\\\\NUnit\\\\nunit-3.0\\\\work\\\\NUnitFramework\\\\src\\\\mock-assembly\\\\MockAssembly.cs:line 121",
                "line": undefined,
                "message": "Intentional failure",
                "path": undefined,
              },
              "name": "FailingTest",
              "result": "failed",
              "time": 23,
            },
          ],
        },
        TestGroupResult {
          "name": "InconclusiveTest",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "InconclusiveTest",
              "result": "success",
              "time": 1,
            },
          ],
        },
        TestGroupResult {
          "name": "MockTest1",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "MockTest1",
              "result": "success",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "MockTest2",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "MockTest2",
              "result": "success",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "MockTest3",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "MockTest3",
              "result": "success",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "MockTest4",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "MockTest4",
              "result": "skipped",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "MockTest5",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "MockTest5",
              "result": "skipped",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "NotRunnableTest",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "NotRunnableTest",
              "result": "skipped",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "TestWithException",
          "tests": Array [
            TestCaseResult {
              "error": Object {
                "details": "   at NUnit.Tests.Assemblies.MockTestFixture.MethodThrowsException() in D:\\\\Dev\\\\NUnit\\\\nunit-3.0\\\\work\\\\NUnitFramework\\\\src\\\\mock-assembly\\\\MockAssembly.cs:line 158
   at NUnit.Tests.Assemblies.MockTestFixture.TestWithException() in D:\\\\Dev\\\\NUnit\\\\nunit-3.0\\\\work\\\\NUnitFramework\\\\src\\\\mock-assembly\\\\MockAssembly.cs:line 153",
                "line": undefined,
                "message": "System.ApplicationException : Intentional Exception",
                "path": undefined,
              },
              "name": "TestWithException",
              "result": "failed",
              "time": 2,
            },
          ],
        },
        TestGroupResult {
          "name": "TestWithManyProperties",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "TestWithManyProperties",
              "result": "success",
              "time": 0,
            },
          ],
        },
      ],
      "indentationDepth": 1,
      "name": "MockTestFixture",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [],
      "indentationDepth": 1,
      "name": "BadFixture",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [],
      "indentationDepth": 1,
      "name": "FixtureWithTestCases",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [
        TestGroupResult {
          "name": "MethodWithParameters(2,2)",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "MethodWithParameters(2,2)",
              "result": "success",
              "time": 6,
            },
          ],
        },
        TestGroupResult {
          "name": "MethodWithParameters(9,11)",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "MethodWithParameters(9,11)",
              "result": "success",
              "time": 0,
            },
          ],
        },
      ],
      "indentationDepth": 2,
      "name": "MethodWithParameters",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [],
      "indentationDepth": 1,
      "name": "IgnoredFixture",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [],
      "indentationDepth": 1,
      "name": "ParameterizedFixture",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [
        TestGroupResult {
          "name": "Test1",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "Test1",
              "result": "success",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "Test2",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "Test2",
              "result": "success",
              "time": 0,
            },
          ],
        },
      ],
      "indentationDepth": 2,
      "name": "ParameterizedFixture(42)",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [
        TestGroupResult {
          "name": "Test1",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "Test1",
              "result": "success",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "Test2",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "Test2",
              "result": "success",
              "time": 0,
            },
          ],
        },
      ],
      "indentationDepth": 2,
      "name": "ParameterizedFixture(5)",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [
        TestGroupResult {
          "name": "TestCase",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "TestCase",
              "result": "success",
              "time": 0,
            },
          ],
        },
      ],
      "indentationDepth": 1,
      "name": "OneTestCase",
      "totalTime": NaN,
    },
    TestSuiteResult {
      "groups": Array [
        TestGroupResult {
          "name": "MyTest",
          "tests": Array [
            TestCaseResult {
              "error": undefined,
              "name": "MyTest",
              "result": "success",
              "time": 1,
            },
          ],
        },
      ],
      "indentationDepth": 1,
      "name": "MockTestFixture",
      "totalTime": NaN,
    },
  ],
  "totalTime": 154,
}
`;
